src/core/regimes.rkt

```diff
@@ -175,127 +175,131 @@
-  ;; prefix of f is for final result vectors
-  ;; a for acost vectors
-  ;; b for candidate index
-  ;; c for alt index
-  ;; d for previous index
-  ;; This is where the high level bulk of the algorithm is applied
-  ;; We get the final splitpoints by applying add-splitpoints as many times as we want
-  (define-values (pa pb pd) (initial))
-  (define-values (fa fb fd)
-    ; short circuit if there is no other alts to consider
-    (if (> num-candidates 1)
-      (let loop ([pa pa] [pb pb] [pd pd])
-      (define-values (na nb nd) (add-splitpoint pa pb pd))
-      (if (equal? nb pb) ;; only need to compare candidate index
-          (values na nb nd)
-          (loop na nb nd)))
-    (values pa pb pd)))
-    
-    ;; From here down is messy code translating from 4 vectors back to
-    ;; the original list of split points
-    (define fixed-final (make-vector num-points))
-    (define fp (list->vector (range 1 (+ num-points 1))))
-    (for ([idx (in-range 0 num-points)])
-      (define a (flvector-ref fa idx))
-      (define b (vector-ref fb idx))
-      (define c (vector-ref fp idx))
-      (define d (vector-ref fd idx))
-      (vector-set! fixed-final idx (cand a b c d)))
+   ;; Save current values for the current point we are working on.
+   (define current-alt-error (flvector-ref result-error-sums point-idx))
+   (define current-alt-idx (vector-ref result-alt-idxs point-idx))
+   (define current-prev-idx (vector-ref result-prev-idxs point-idx))
+   ;; We have now have the index of the best alt and it's error up to our 
+   ;; current point-idx.
+   ;; Now we compare against our current best saved in the 3 vectors above
+   (for ([prev-split-idx (in-range 0 point-idx)])
+    (when (vector-ref can-split-vec (+ prev-split-idx 1))
+     ;; Re compute the error sum for a potential better alt
+     (define alt-error-sum (fl+ (fl+ (flvector-ref result-error-sums prev-split-idx)
+                    (vector-ref best-alt-cost prev-split-idx)) min-weight))
+     ;; pre-compute values for tie breaking
+     (define adjusted-cost current-alt-error)
+     (define current-best-alt-idx (vector-ref best-alt-idx prev-split-idx))
+     ;; Check if the new alt-error-sum is better then the current
+     (define set-cond
+      ;; give benefit to previous best alt
+      (cond [(fl< alt-error-sum adjusted-cost) #t]
+            ;; Tie breaker if error are the same favor first alt
+            [(and (fl= alt-error-sum adjusted-cost)
+                  (> current-alt-idx current-best-alt-idx)) #t]
+            ;; Tie breaker for if error and alt is the same
+            [(and (fl= alt-error-sum adjusted-cost)
+                  (= current-alt-idx current-best-alt-idx)
+                  (> current-prev-idx prev-split-idx)) #t]
+            [else #f]))
+      (when set-cond
+       (set! current-alt-error alt-error-sum)
+       (set! current-alt-idx (vector-ref best-alt-idx prev-split-idx))
+       (set! current-prev-idx prev-split-idx))))
```

Comment:
Now that it all works I wonder if we still need the complex tie breaking condition. Probably not?

Reply:
I feel like it would be better to leave on even thought it's slower? My 1st thought is using something like `#if DEBUG` so that when testing the code there is an assert for correctness but release there isn't maybe that bad practice as release is different than debug.

Maybe the ordering could just be left as a simple comment?

Reply:
Let's leave it in for now and see how much of a speedup all the other stuff gives.

Reply:
Aside: Ci fails on `quadp` if I remove this.

Reply:
I guess that answers that question!